plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/s3/S3FileSystemExchangeStorage.java

```diff
@@ -0,0 +1,564 @@
+    }
+
+    @Override
+    public long size(URI uri, Optional<SecretKey> secretKey)
+            throws IOException
+    {
+        checkArgument(!isDirectory(uri), "expected a file URI but got a directory URI");
+        HeadObjectResponse response = getHeadObjectResponse(uri, secretKey);
+        if (response == null) {
+            throw new FileNotFoundException("File does not exist: " + uri);
+        }
+        return response.contentLength();
+    }
+
+    @Override
+    public int getWriteBufferSizeInBytes()
+    {
+        return multiUploadPartSize;
+    }
+
+    @Override
+    public void close()
+            throws IOException
+    {
+        try (Closer closer = Closer.create()) {
+            closer.register(deleteExecutor::shutdown);
+            closer.register(s3Client::close);
+            closer.register(s3AsyncClient::close);
+        }
+    }
+
+    private void deleteRecursivelySync(URI uri)
+    {
+        if (isDirectory(uri)) {
+            listDirectories(uri).forEach(this::deleteRecursively);
+            listFiles(uri).forEach(file -> deleteObject(getBucketName(file), keyFromUri(file)));
+            deleteObject(getBucketName(uri), keyFromUri(uri) + DIRECTORY_SUFFIX);
+        }
+        else {
+            deleteObject(getBucketName(uri), keyFromUri(uri));
+        }
+    }
+
+    private HeadObjectResponse getHeadObjectResponse(URI uri, Optional<SecretKey> secretKey)
+            throws IOException
+    {
+        HeadObjectRequest.Builder headObjectRequestBuilder = HeadObjectRequest.builder()
+                .bucket(getBucketName(uri))
+                .key(keyFromUri(uri));
+        S3RequestUtil.configureEncryption(secretKey, headObjectRequestBuilder);
+
+        try {
+            return s3Client.headObject(headObjectRequestBuilder.build());
+        }
+        catch (AwsServiceException e) {
+            if (e instanceof NoSuchKeyException) {
+                return null;
+            }
+            throw new IOException(e);
+        }
+    }
+
+    private ListObjectsV2Response listS3Directory(URI dir)
+    {
+        String key = keyFromUri(dir);
+        if (!key.isEmpty()) {
+            key += PATH_SEPARATOR;
+        }
+
+        ListObjectsV2Request request = ListObjectsV2Request.builder()
+                .bucket(getBucketName(dir))
+                .prefix(key)
+                .delimiter(PATH_SEPARATOR)
+                .build();
+
+        return s3Client.listObjectsV2(request);
+    }
+
+    private void deleteObject(String bucketName, String key)
+    {
+        DeleteObjectRequest request = DeleteObjectRequest.builder()
+                .bucket(bucketName)
+                .key(key)
+                .build();
+        s3Client.deleteObject(request);
+    }
+
+    private static String getBucketName(URI uri)
+    {
+        if (uri.getHost() != null) {
+            return uri.getHost();
+        }
+
+        if (uri.getUserInfo() == null) {
+            return uri.getAuthority();
+        }
```

Comment:
I do not understand logic here. Can you provide a comment to explain what is happening here?

Reply:
The function is copied from `TrinoS3FileSyste`, see https://github.com/trinodb/trino/blob/master/plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java#L1711-L1719. Let me copy the comments too.